integration/port_forwarding_test.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/session"
+	"github.com/gravitational/teleport/lib/srv/regular"
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+)
+
+func extractPort(svr *httptest.Server) (int, error) {
+	u, err := url.Parse(svr.URL)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	n, err := strconv.Atoi(u.Port())
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	return n, nil
+}
+
+func waitForSessionToBeEstablishedWithContext(ctx context.Context, namespace string, site auth.ClientI) ([]session.Session, error) {
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+
+		case <-ticker.C:
+			ss, err := site.GetSessions(namespace)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			if len(ss) > 0 {
+				return ss, nil
+			}
```

Comment:
I recently learned about [require.Eventually](https://pkg.go.dev/github.com/stretchr/testify/require#Eventually), could be useful for cases like this

Reply:
Neat. Unfortunately in this case I need to wait on an arbtrary context rather than a duration. Buut good to know it exists :-)